environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/AbstractEnvironmentCreationAction.java

```diff
@@ -0,0 +1,45 @@
+package com.sequenceiq.environment.environment.flow.creation;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.statemachine.StateContext;
+
+import com.sequenceiq.cloudbreak.common.event.ResourceCrnPayload;
+import com.sequenceiq.cloudbreak.logger.MdcContext;
+import com.sequenceiq.environment.environment.flow.creation.event.EnvCreationStateSelectors;
+import com.sequenceiq.flow.core.AbstractAction;
+import com.sequenceiq.flow.core.CommonContext;
+import com.sequenceiq.flow.core.FlowParameters;
+
+abstract class AbstractEnvironmentCreationAction<P extends ResourceCrnPayload>
+        extends AbstractAction<EnvCreationState, EnvCreationStateSelectors, CommonContext, P> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEnvironmentCreationAction.class);
+
+    protected AbstractEnvironmentCreationAction(Class<P> payloadClass) {
+        super(payloadClass);
+    }
+
+    @Override
+    protected CommonContext createFlowContext(FlowParameters flowParameters, StateContext<EnvCreationState, EnvCreationStateSelectors> stateContext,
+            P payload) {
+        return new CommonContext(flowParameters);
+    }
+
+    @Override
+    protected Object getFailurePayload(P payload, Optional<CommonContext> flowContext, Exception ex) {
+        return payload;
```

Comment:
Can we have an additional log line here? I think it's worth it :slightly_smiling_face: